src/client-side-encryption/autoEncrypter.ts

```diff
@@ -0,0 +1,641 @@
+    }
+
+    mongoCryptOptions.kmsProviders = !Buffer.isBuffer(this._kmsProviders)
+      ? (serialize(this._kmsProviders) as Buffer)
+      : this._kmsProviders;
+
+    if (options.options?.logger) {
+      mongoCryptOptions.logger = options.options.logger;
+    }
+
+    if (options.extraOptions && options.extraOptions.cryptSharedLibPath) {
+      mongoCryptOptions.cryptSharedLibPath = options.extraOptions.cryptSharedLibPath;
+    }
+
+    if (options.bypassQueryAnalysis) {
+      mongoCryptOptions.bypassQueryAnalysis = options.bypassQueryAnalysis;
+    }
+
+    this._bypassMongocryptdAndCryptShared = this._bypassEncryption || !!options.bypassQueryAnalysis;
+
+    if (options.extraOptions && options.extraOptions.cryptSharedLibSearchPaths) {
+      // Only for driver testing
+      mongoCryptOptions.cryptSharedLibSearchPaths = options.extraOptions.cryptSharedLibSearchPaths;
+    } else if (!this._bypassMongocryptdAndCryptShared) {
+      mongoCryptOptions.cryptSharedLibSearchPaths = ['$SYSTEM'];
+    }
+
+    Object.assign(mongoCryptOptions, { cryptoCallbacks });
+    const MongoCrypt = AutoEncrypter.getMongoCrypt();
+    this._mongocrypt = new MongoCrypt(mongoCryptOptions);
+    this._contextCounter = 0;
+
+    if (
+      options.extraOptions &&
+      options.extraOptions.cryptSharedLibRequired &&
+      !this.cryptSharedLibVersionInfo
+    ) {
+      throw new MongoError('`cryptSharedLibRequired` set but no crypt_shared library loaded');
+    }
+
+    // Only instantiate mongocryptd manager/client once we know for sure
+    // that we are not using the CSFLE shared library.
+    if (!this._bypassMongocryptdAndCryptShared && !this.cryptSharedLibVersionInfo) {
+      this._mongocryptdManager = new MongocryptdManager(options.extraOptions);
+      const clientOptions: MongoClientOptions = {
+        serverSelectionTimeoutMS: 10000
+      };
+
+      if (options.extraOptions == null || typeof options.extraOptions.mongocryptdURI !== 'string') {
+        clientOptions.family = 4;
+      }
+
+      this._mongocryptdClient = new MongoClient(this._mongocryptdManager.uri, clientOptions);
+    }
+  }
+
+  /**
+   * Initializes the auto encrypter by spawning a mongocryptd and connecting to it.
+   *
+   * This function is a no-op when bypassSpawn is set or the crypt shared library is used.
+   */
+  init(callback: Callback<MongoClient>) {
+    if (this._bypassMongocryptdAndCryptShared || this.cryptSharedLibVersionInfo) {
+      return callback();
+    }
+    if (!this._mongocryptdManager) {
+      return callback(
+        new MongoRuntimeError(
+          'Reached impossible state: mongocryptdManager is undefined when neither bypassSpawn nor the shared lib are specified.'
+        )
+      );
+    }
+    if (!this._mongocryptdClient) {
+      return callback(
+        new MongoRuntimeError(
+          'Reached impossible state: mongocryptdClient is undefined when neither bypassSpawn nor the shared lib are specified.'
+        )
+      );
+    }
```

Comment:
Typescript complains here, because we (developers) know that `init` is a no-op when `this._bypassMongocryptdAndCryptShared || this.cryptSharedLibVersionInfo` is specified and that if neither of those are true, we always have a _mongocryptdManager` and client.  We can cast it away but I think errors are a good practice, because these provide actionable feedback if the assumption that these properties are always defined when we need to spawn a mongocryptd is ever violated.

Reply:
Maybe a `MongoCryptImpossibleState` error?

Reply:
it isn't fixable for users if they encounter this error - we create a `mongocryptd` client so if they encounter this error, we have a bug